cf-check/backup.c

```diff
@@ -0,0 +1,198 @@
+
+int backup_main(int argc, char **argv)
+{
+    printf("backup only implemented for LMDB.\n");
+    return 1;
+}
+
+#else
+
+#include <stdio.h>
+
+bool copy_file(const char *src, const char *dst)
+{
+    assert(src != NULL);
+    assert(dst != NULL);
+
+    printf("Copying: '%s' -> '%s'\n", src, dst);
+    FILE *in = fopen(src, "r");
+    if (in == NULL)
+    {
+        return false;
+    }
+    FILE *out = fopen(dst, "w");
+    if (out == NULL)
+    {
+        fclose(in);
+        return false;
+    }
+    size_t read = 0;
+    size_t written = 0;
+    bool ret = true;
+    do
+    {
+        const size_t bufsize = 1024;
+        char buf[1024] = { 0 };
+
+        read = fread(buf, sizeof(char), bufsize, in);
+
+        written = fwrite(buf, sizeof(char), read, out);
+        while ((!ferror(out)) && (written < read))
+        {
+            written += fwrite(buf + written, sizeof(char), read - written, out);
+        }
+
+        if (read != written || ferror(in) || ferror(out))
+        {
+            break;
+        }
+        if (feof(in))
+        {
+            read = 0;
+        }
+    } while (read > 0);
+
+    if (ferror(in))
+    {
+        printf("Error encountered while reading '%s'\n", src);
+        ret = false;
+    }
+    else if (ferror(out))
+    {
+        printf("Error encountered while writing '%s'\n", dst);
+        ret = false;
+    }
+    else if (read != 0 && read != written)
+    {
+        printf("Did not copy the whole file\n");
+        ret = false;
+    }
+
+    fclose(in);
+    fclose(out);
+    return ret;
+}
+
+const char *filename_part(const char *path)
+{
+    assert(path != NULL);
+
+    const char *filename = strrchr(path, '/');
+
+    if (filename != NULL)
+    {
+        filename += 1;
+    }
+    else
+    {
+        filename = path;
+    }
+
+    if (filename[0] == '\0')
+    {
+        return NULL;
+    }
+
+    return filename;
+}
+
+bool copy_file_to_folder(const char *src, const char *dst_dir)
+{
+    assert(src != NULL);
+    assert(dst_dir != NULL);
+
+    const char *filename = filename_part(src);
+    if (filename == NULL)
+    {
+        printf("Cannot find filename in '%s'\n", src);
+        return false;
+    }
+
+    char dst[PATH_MAX] = { 0 };
+    const int s = snprintf(dst, PATH_MAX, "%s%s", dst_dir, filename);
+    if (s >= PATH_MAX)
+    {
+        printf("Backup destination path too long: '%s...'\n", dst);
+        return false;
+    }
+
+    if (!copy_file(src, dst))
+    {
+        printf("Copying '%s' failed\n", filename);
+        return false;
+    }
+
+    return true;
+}
+
+const char *create_backup_dir()
+{
+    const char *const backup_root = "/var/cfengine/backup/";
+    if (mkdir(backup_root, 0700) != 0)
+    {
+        if (errno != EEXIST)
+        {
+            printf("Could not create directory '%s' (%s)\n", backup_root, strerror(errno));
+            return NULL;
+        }
+    }
+    static char backup_dir[PATH_MAX];
+    time_t ts = time(NULL);
+
```

Comment:
reading glib to see when time() returns -1 was fun. Didn't quite get all the way through it! ;) Probably OK to not check for -1 or errors?

Reply:
I will add a check for errors here also.